### PR TITLE
Fixes #1162 - Consume api pushdown of resource descriptions

### LINKED ISSUE #1162: Consume new IResourceSnapshot API in place of ResourceDescription
platform.core has a new IResourceSnapshot API which replaces the ResourceDescription classes. This should be consumed by ui.ide if at all possible to clean up and remove old code and standardize on the new classes.

### PR DESCRIPTION
**Pull request by robstryker**

### CODE REVIEW

**Comment by robstryker** [comment]
May make a few revisions here, so don't look to merge until I'm confident in the result.

**Comment by mickaelistria** [comment]
I'm marking it as draft then. Please switch it as "ready for review" when ready for review.

File: bundles/org.eclipse.ui.ide/src/org/eclipse/ui/ide/undo/ResourceDescription.java [hunk 1]
```diff
@@ -31,69 +27,100 @@
  * This class is not intended to be extended by clients.
  *
  * @since 3.3
- *
+ * @deprecated
```
Comment (mickaelistria) [inline bundles/org.eclipse.ui.ide/src/org/eclipse/ui/ide/undo/ResourceDescription.java]: Please add a note to this annotation telling which class should be used instead.

**Review by mickaelistria — CHANGES_REQUESTED** [review]

**Review by mickaelistria — APPROVED** [review]
We'll merge it by the end of week, after M1 is done.

File: bundles/org.eclipse.ui.ide/META-INF/MANIFEST.MF [hunk 2]
```diff
@@ -2,7 +2,7 @@ Manifest-Version: 1.0
 Bundle-ManifestVersion: 2
 Bundle-Name: %Plugin.name
 Bundle-SymbolicName: org.eclipse.ui.ide; singleton:=true
-Bundle-Version: 3.21.200.qualifier
+Bundle-Version: 3.22.0.qualifier
```
Comment (szarnekow) [inline bundles/org.eclipse.ui.ide/META-INF/MANIFEST.MF:5]: If we use newer API now, the version constraint on core resource should be changed, too, no?

**Review by szarnekow — CHANGES_REQUESTED** [review]

File: bundles/org.eclipse.ui.ide/META-INF/MANIFEST.MF (hunk 2 above)
Comment (iloveeclipse) [reply bundles/org.eclipse.ui.ide/META-INF/MANIFEST.MF:5]: Good point, but why do we change **minor** version segment in first place? Shouldn't that be 3.21.300?

Comment (iloveeclipse) [reply bundles/org.eclipse.ui.ide/META-INF/MANIFEST.MF:5]: Ok, I see, PR doesn't explain that. "Consume" sounds like "use it" but in fact, the public API in several classes has changed...

Comment (mickaelistria) [reply bundles/org.eclipse.ui.ide/META-INF/MANIFEST.MF:5]: Thanks for the reminder @szarnekow ! @robstryker can you please restrict the minimal version for the requirement on core.resources to ensure we link to a version that has the new API ?

Comment (robstryker) [reply bundles/org.eclipse.ui.ide/META-INF/MANIFEST.MF:5]: Done

**Comment by robstryker** [comment]
Rebased and ready once build is done

File: bundles/org.eclipse.ui.ide/src/org/eclipse/ui/internal/ide/undo/FolderDescription.java [hunk 3]
```diff
@@ -1,104 +0,0 @@
...
-
-	@Override
-	public void createExistentResourceFromHandle(IResource resource, IProgressMonitor mon) throws CoreException {
-		Assert.isLegal(resource instanceof IFolder);
-		if (resource.exists()) {
-			return;
-		}
-		IFolder folderHandle = (IFolder) resource;
-		SubMonitor subMonitor = SubMonitor.convert(mon, 300);
-		subMonitor.setTaskName(UndoMessages.FolderDescription_NewFolderProgress);
```
Comment (szarnekow) [inline bundles/org.eclipse.ui.ide/src/org/eclipse/ui/internal/ide/undo/FolderDescription.java:88]: Do we have obsolete keys in UndoMessages, now?

File: bundles/org.eclipse.ui.ide/src/org/eclipse/ui/ide/undo/AbstractCreateResourcesOperation.java [hunk 4]
```diff
@@ -80,7 +81,7 @@ protected boolean updateResourceChangeDescriptionFactory(
 				modified = true;
 			}
 		} else {
-			for (ResourceDescription resourceDescription : resourceDescriptions) {
+			for (IResourceSnapshot resourceDescription : resourceDescriptions) {
```
Comment (szarnekow) [inline bundles/org.eclipse.ui.ide/src/org/eclipse/ui/ide/undo/AbstractCreateResourcesOperation.java]: Look at line 48: Is this a rawtype reference here? Does it need to be a rawtype?

**Comment by robstryker** [comment]
Build seems to have timed out @mickaelistria

File: bundles/org.eclipse.ui.ide/src/org/eclipse/ui/ide/undo/AbstractCreateResourcesOperation.java (hunk 4 above)
Comment (robstryker) [reply bundles/org.eclipse.ui.ide/src/org/eclipse/ui/ide/undo/AbstractCreateResourcesOperation.java]: Unchecked references to IResourceSnapshot and its sub-interfaces is rampant throughout this patch. 

The biggest problem I've run into when trying to add types to all of these is that much of the current code makes use of arrays. And I am unable to instantiate a new array with a generic type attached. 

Example:

```
	public CreateProjectOperation(IProjectDescription projectDescription, String label) {
		super(new IResourceSnapshot[] { ResourceSnapshotFactory.fromProjectDescription(projectDescription) }, label);
	}
```

The super constructor accepts `IResourceSnapshot<? extends IResource>[]` as the first parameter. But if I try to add generics to this method as follows, it won't compile:

`super(new IResourceSnapshot<IProject>[] { ResourceSnapshotFactory.fromProjectDescription(projectDescription) }, label);`  

The above line shows the error `Cannot create a generic array of IResourceSnapshot<IProject>`

And this happens all over the codebase if I try to make sure it's all typesafe.

Comment (robstryker) [reply bundles/org.eclipse.ui.ide/src/org/eclipse/ui/ide/undo/AbstractCreateResourcesOperation.java]: To be honest I originally designed this API without generics, but now that @mickaelistria  has added generics to it, I struggle to clean up this code in an appropriate way without having to make significant changes.

**Comment by akurtakov** [comment]
I've retriggered the build. If you want to retrigger just amend the build and push again and that will cause the verification build to rerun.

**Comment by mickaelistria** [comment]
Arrays and generics don't play well together. If you want to allow better typing, you'll need to switch to List.

**Comment by robstryker** [comment]
I've explained one of my main issues here:   https://github.com/eclipse-platform/eclipse.platform/pull/729

This pr on platform core will likely fix the issue.

**Comment by robstryker** [comment]
I've updated this PR to now depend on the platform modification to the new API. 

I feel the platform cleanup will resolve all of the issues I have and make this PR ready to push once platform has the changes available.

**Comment by robstryker** [comment]
Forgot to link to the platform core PR:   https://github.com/eclipse-platform/eclipse.platform/pull/729/files

**Comment by mickaelistria** [comment]
We have to wait for next I-Build before the "ontinuous-integration/jenkins/pr-head" reports according to merged https://github.com/eclipse-platform/eclipse.platform/pull/729

**Comment by mickaelistria** [comment]
Build reports an API issue:
```
1 API ERRORS
[API ERROR] File MANIFEST.MF at line 5: The major version should be incremented in version 3.22.0, since API breakage occurred since version 3.21.100 - Superinterface that has methods to be implemented is added to org.eclipse.ui.ide.undo.ResourceDescription (location: /home/jenkins/agent/workspace/eclipse.platform.ui_PR-1163/bundles/org.eclipse.ui.ide/META-INF/MANIFEST.MF)
```

**Comment by mickaelistria** [comment]
Thank you!

**Pull request merged**